### PR TITLE
ARROW-12060: [Python] Enable calling compute functions on Expressions

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
This makes the following already possible (to use in filter or projection expression in dataset):

```
In [1]: import pyarrow.compute as pc

In [2]: import pyarrow.dataset as ds

In [3]: pc.hour(ds.field("datetime_col"))
Out[3]: <pyarrow.dataset.Expression hour(datetime_col)>

In [4]: pc.days_between(ds.field("col1"), ds.field("col2"))
Out[4]: <pyarrow.dataset.Expression days_between(col1, col2)>
```

File: python/pyarrow/_compute.pyx [hunk 1]
```diff
@@ -2019,6 +2019,25 @@ cdef class Expression(_Weakrefable):
...
+            shared_ptr[CFunctionOptions] c_options=(
+                <shared_ptr[CFunctionOptions]> nullptr)
+
+        for argument in arguments:
+            if not isinstance(argument, Expression):
+                raise TypeError("only other expressions allowed as arguments")
+            c_arguments.push_back((<Expression> argument).expr)
+
+        # if options is not None:
+        #     c_options = make_shared[CFunctionOptions](options.get_options())
```
Comment (jorisvandenbossche) [inline python/pyarrow/_compute.pyx:2036]: Given a python `FunctionOptions` class object, I need to get a shared pointer to the C++ object (`get_options()` returns a `const CFunctionOptions*`, but `CmakeCallExpressions` needs a `shared_ptr[CFunctionOptions]`). 

But I haven't directly found how to do that conversion (cc @pitrou)

Comment (pitrou) [reply python/pyarrow/_compute.pyx:2036]: You should probably replace the `unique_ptr` in the FunctionOptions declaration with a `shared_ptr`.

Comment (jorisvandenbossche) [reply python/pyarrow/_compute.pyx:2036]: The  `DeserializeFunctionOptions` in C++ (in function_internal.h, and several other methods there) returns a `unique_ptr`(from https://github.com/apache/arrow/pull/10511). Can that be changed there as well?

Comment (pitrou) [reply python/pyarrow/_compute.pyx:2036]: You can, but I don't think it's necessary on the C++ side. On the Python side, you can convert the unique_ptr into a shared_ptr (by moving it).

Comment (jorisvandenbossche) [reply python/pyarrow/_compute.pyx:2036]: Yes, that's what I did in the meantime. The only difficulty I ran into is that moving a uniqe_ptr to shared_ptr doesn't seem to work in cython, but with our helper `to_shared` it worked.

File: python/pyarrow/_compute.pyx [hunk 2]
```diff
@@ -2020,17 +2025,22 @@ cdef class Expression(_Weakrefable):
...
 
     @staticmethod
-    cdef Expression _call(str function_name, list arguments,
-                          shared_ptr[CFunctionOptions] options=(
-                              <shared_ptr[CFunctionOptions]> nullptr)):
+    def _call(str function_name, list arguments, FunctionOptions options=None):
         cdef:
             vector[CExpression] c_arguments
+            shared_ptr[CFunctionOptions] c_options=(
+                <shared_ptr[CFunctionOptions]> nullptr)
```
Comment (pitrou) [inline python/pyarrow/_compute.pyx:2032]: You don't need to explicitly initialize this, do you?

**Review by pitrou — APPROVED** [review]
Neat!

File: python/pyarrow/_compute.pyx (hunk 2 above)
Comment (jorisvandenbossche) [reply python/pyarrow/_compute.pyx:2032]: Ah, yes, passing an unitialized `c_options` to CMakeCallExpression seems to work as well

**Pull request closed without merging**